plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,245 @@
+	}
+	// Avro doesn't have a Tag/Field distinction, so we have to tell
+	// Telegraf which items are our tags.
+	for _, tag := range p.Tags {
+		sTag, err := internal.ToString(data[tag])
+		if err != nil {
+			p.Log.Warnf("Could not convert %v to string for tag %q: %v", data[tag], tag, err)
+			continue
+		}
+		tags[tag] = sTag
+	}
+	var fieldList []string
+	if len(p.Fields) != 0 {
+		// If you have specified your fields in the config, you
+		// get what you asked for.
+		fieldList = p.Fields
+	} else {
+		for k := range data {
+			// Otherwise, that which is not a tag is a field
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	// We need to flatten out our fields.  The default (the separator
+	// string is empty) is equivalent to what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	for _, fld := range fieldList {
+		candidate := make(map[string]interface{})
+		candidate[fld] = data[fld] // 1-item map
+		flat, err := flatten.Flatten(candidate, "", sep)
+		if err != nil {
+			return nil, fmt.Errorf("flatten field %q failed: %w", fld, err)
+		}
+		for k, v := range flat {
+			fields[k] = v
+		}
+	}
+
+	var schemaObj map[string]interface{}
+	if err := json.Unmarshal([]byte(schema), &schemaObj); err != nil {
+		return nil, fmt.Errorf("unmarshaling schema failed: %w", err)
+	}
+	if len(fields) == 0 {
+		// A telegraf metric needs at least one field.
+		return nil, errors.New("number of fields is 0; unable to create metric")
+	}
+	// Now some fancy stuff to extract the measurement.
+	// If it's set in the configuration, use that.
+	name := p.Measurement
+	separator := "."
+	if name == "" {
+		// Try using the namespace defined in the schema. In case there
+		// is none, just use the schema's name definition.
+		nsStr, ok := schemaObj["namespace"].(string)
+		// namespace is optional
+		if !ok {
+			separator = ""
+		}
+
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + separator + nStr
+	}
+	// Still don't have a name?  Guess we should use the metric name if
+	// it's set.
+	if name == "" {
+		name = p.MetricName
+	}
+	// Nothing?  Give up.
+	if name == "" {
+		return nil, errors.New("could not determine measurement name")
+	}
+	timestamp := time.Now()
+	if p.Timestamp != "" {
+		rawTime := fields[p.Timestamp]
+		timeIncs, ok := rawTime.(int64)
+		if !ok {
+			return nil, fmt.Errorf("timestamp %v could not be converted to int64", rawTime)
+		}
+		if p.TimestampFormat == "unix_us" {
+			timestamp = time.UnixMicro(timeIncs)
+		} else if p.TimestampFormat == "unix_ms" {
+			timestamp = time.UnixMilli(timeIncs)
+		} else if p.TimestampFormat == "unix_ns" {
+			// If it's a single Avro field, it's less than
+			// 2**63 - 1 ... which will last a while
+			timestamp = time.Unix(0, timeIncs)
+		} else {
+			timestamp = time.Unix(timeIncs, 0)
+		}
```

Comment:
We have a helper to do this check out `timestamp, err := internal.ParseTimestamp(p.TimestampFormat, rawtime, "")`. I think it would be best to send in the timestamp as a string and you will need to import internal as well: